server/src/main/java/org/elasticsearch/cluster/routing/allocation/ShardsAvailabilityHealthIndicatorService.java

```diff
@@ -146,7 +255,133 @@ private static boolean isUnassignedDueToNewInitialization(ShardRouting routing)
+            if (LOGGER.isTraceEnabled()) {
+                LOGGER.trace(
+                    "[{}]: Working with decisions: [{}]",
+                    shardRouting.shardId(),
+                    allocateDecision.getNodeDecisions()
+                        .stream()
+                        .map(
+                            n -> n.getCanAllocateDecision()
+                                .getDecisions()
+                                .stream()
+                                .map(d -> d.label() + ": " + d.type())
+                                .collect(Collectors.toList())
+                        )
+                        .collect(Collectors.toList())
+                );
+            }
+            List<NodeAllocationResult> nodeAllocationResults = allocateDecision.getNodeDecisions();
+            boolean diagnosisFound = checkIsAllocationDisabled(diagnosisOutput, shardRouting, nodeAllocationResults);
+            IndexMetadata index = state.metadata().index(shardRouting.index());
+            if (index != null) {
+                diagnosisFound |= checkDataTierRelatedIssues(diagnosisOutput, index, shardRouting, nodeAllocationResults);
+            }
+            if (diagnosisFound == false) {
+                diagnosisOutput.accept(ACTION_CHECK_ALLOCATION_EXPLAIN_API, shardRouting);
+            }
+        }
+    }
+
+    private static Predicate<NodeAllocationResult> nodeHasDeciderResult(String deciderName, Decision.Type outcome) {
+        return (nodeResult) -> nodeResult.getCanAllocateDecision()
+            .getDecisions()
+            .stream()
+            .anyMatch(decision -> deciderName.equals(decision.label()) && outcome == decision.type());
+    }
+
+    private boolean checkIsAllocationDisabled(
+        BiConsumer<UserAction.Definition, ShardRouting> diagnosisOutput,
+        ShardRouting shardRouting,
+        List<NodeAllocationResult> nodeAllocationResults
+    ) {
+        if (nodeAllocationResults.stream().allMatch(nodeHasDeciderResult(EnableAllocationDecider.NAME, Decision.Type.NO))) {
+            diagnosisOutput.accept(ACTION_ENABLE_ALLOCATIONS, shardRouting);
+            return true;
+        }
+        return false;
+    }
+
+    private boolean checkDataTierRelatedIssues(
+        BiConsumer<UserAction.Definition, ShardRouting> diagnosisOutput,
+        IndexMetadata indexMetadata,
+        ShardRouting shardRouting,
+        List<NodeAllocationResult> nodeAllocationResults
+    ) {
+        boolean diagnosisFound = false;
+        if (indexMetadata.getTierPreference().size() > 0) {
+            List<NodeAllocationResult> dataTierNodes = nodeAllocationResults.stream()
+                .filter(nodeHasDeciderResult(DATA_TIER_ALLOCATION_DECIDER_NAME, Decision.Type.YES))
+                .collect(Collectors.toList());
+            if (dataTierNodes.isEmpty()) {
+                diagnosisOutput.accept(ACTION_ENABLE_TIERS, shardRouting);
+                diagnosisFound = true;
```

Comment:
If none of the DiscoveryNodes are marked with any tier roles, I don't think that method will return any tiers. This case is likely to be rare since a combination of the `data` role is acceptable to allocate an index from any tier, not to mention tier preference accepts a list of tier roles specifically to avoid this problem if an entire tier is missing.

I think with all that in mind it would be safe to simply iterate over the preferred tiers list on the index and mark each tier found as needing to be enabled on the cluster.

Reply:
Applied in [db033ef](https://github.com/elastic/elasticsearch/pull/85071/commits/db033efcf8f0c16905bdba2162edaa785a645632)